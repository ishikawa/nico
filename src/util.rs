use std::cell::RefCell;
use std::rc::Rc;

pub fn wrap<T>(ty: T) -> Rc<RefCell<T>> {
    Rc::new(RefCell::new(ty))
}

pub trait UniqueNaming {
    fn next(&mut self) -> String;
}

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug)]
pub struct SequenceNaming {
    type_var_index: usize,
    prefix: String,
}

impl SequenceNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }
}

impl UniqueNaming for SequenceNaming {
    fn next(&mut self) -> String {
        let i = self.type_var_index;
        self.type_var_index += 1;
        format!("{}{}", self.prefix, i)
    }
}
