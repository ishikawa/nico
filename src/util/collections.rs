use crate::arena::BumpaloVec;
use std::cell::Ref;

#[derive(Debug)]
pub struct RefVecIter<'r, 'arena, T> {
    vec: Ref<'r, BumpaloVec<'arena, T>>,
}

impl<'r, 'arena, T> RefVecIter<'r, 'arena, T> {
    pub fn new(vec: Ref<'r, BumpaloVec<'arena, T>>) -> Self {
        Self { vec }
    }
}

impl<'a: 'r, 'r, 'arena, T> IntoIterator for &'a RefVecIter<'r, 'arena, T> {
    type IntoIter = std::slice::Iter<'r, T>;
    type Item = &'r T;

    fn into_iter(self) -> Self::IntoIter {
        self.vec.iter()
    }
}
