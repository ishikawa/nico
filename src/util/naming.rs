use std::collections::HashMap;

/// Generates a unique string while continuously incrementing the index internally.
/// The user can specify a prefix for the generated string.
#[derive(Debug, Default)]
pub struct SequenceNaming {
    names: HashMap<String, i32>,
}

impl SequenceNaming {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next<S: AsRef<str>>(&mut self, name: S) -> String {
        let name = name.as_ref();
        let n = self.names.entry(name.to_string()).or_insert(0);
        let next = format!("{}.{}", name, n);

        *n += 1;
        next
    }
}

/// This struct produces a unique name while giving a sequential
/// number to the name passed to the next() method.
#[derive(Debug)]
pub struct PrefixNaming {
    type_var_index: usize,
    prefix: String,
}

impl PrefixNaming {
    pub fn new(prefix: &str) -> Self {
        Self {
            type_var_index: 0,
            prefix: prefix.to_string(),
        }
    }

    pub fn next(&mut self) -> String {
        let i = self.type_var_index;
        let next = format!("{}{}", self.prefix, i);

        self.type_var_index += 1;
        next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prefix_naming() {
        let mut naming = PrefixNaming::new("my_");

        assert_eq!(naming.next(), "my_0".to_string());
        assert_eq!(naming.next(), "my_1".to_string());
    }

    #[test]
    fn sequence_naming() {
        let mut naming = SequenceNaming::new();

        assert_eq!(naming.next("x"), "x.0".to_string());
        assert_eq!(naming.next("x"), "x.1".to_string());
        assert_eq!(naming.next("y"), "y.0".to_string());
    }
}
