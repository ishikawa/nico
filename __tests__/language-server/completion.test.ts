import { LanguageServer, spawn_server, LanguageServerAgent, Position } from "../util/lsp";
import { filterTestCases, TestCaseBase, readTestFileSync, getTestName } from "../util/testcase";

interface TestCase extends TestCaseBase {
  requests: Array<{ position: Position }>;
}

let server: LanguageServer | undefined;

beforeAll(async () => {
  server = await spawn_server({ completion: true });
});

afterAll(() => {
  if (server) {
    return server.stop();
  }
});

let cases: TestCase[] = [
  // suggests built-in functions
  {
    input: ["p"].join("\n"),
    requests: [
      // p|
      //  ^
      {
        position: {
          line: 0,
          character: 1
        }
      }
    ]
  }
];

filterTestCases(cases).forEach((testCase, i) => {
  let src = readTestFileSync(testCase);
  let name = getTestName(testCase);

  testCase.requests.forEach(({ position }, j) => {
    test(`${i}: completion - \`${name}\` at ${position}`, async done => {
      const agent = new LanguageServerAgent(server!, { sequence: [i, j] });

      // Allow compilation errors
      const _ = await agent.openDocument(name, src);

      const response = await agent.completion(name, position);
      expect(response).toMatchSnapshot();

      done();
    });
  });
});
