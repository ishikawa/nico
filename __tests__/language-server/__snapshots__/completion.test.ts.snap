// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0: completion - \`p\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1000,
  "jsonrpc": "2.0",
  "result": Object {
    "isIncomplete": false,
    "items": Array [
      Object {
        "kind": 3,
        "label": "println_i32",
      },
      Object {
        "kind": 3,
        "label": "println_str",
      },
    ],
  },
}
`;

exports[`1: completion - \`struct Rectangle {
  width: i32
  height: i32
}
let rect = Rectangle { width: 30, height: 50 }

rc\` at [object Object] 1`] = `
Object {
  "error": null,
  "id": 1010,
  "jsonrpc": "2.0",
  "result": Object {
    "isIncomplete": false,
    "items": Array [
      Object {
        "kind": 6,
        "label": "rect",
      },
      Object {
        "kind": 22,
        "label": "Rectangle",
      },
    ],
  },
}
`;
